### ISSUE TITLE
Refresh instead of purge, stale-while-revalidate

### ISSUE THREAD

**Issue by dkarlovi** [darkweak/souin#276]
When doing a `PURGE` by ykey, you might purge a lot of stuff at the same time, instantly creating a cache avalanche / cache slam.

A typical solution is to push changes into the cache instead of cache pulling them on demand. This ensures a single write is done into the cache, and the clients start getting the new version immediately. The issue here is knowing how to write to the underlying system, which complicates things.

Alternative could be [`stale-while-revalidate`](https://www.rfc-editor.org/rfc/rfc5861#section-3) to support a refresh instead of a direct purge.

Basically, this means the cache item is not flushed, it's marked as stale, but a request is still passed through to allow it be refreshed.

Something like

1. `GET /`, `Cache-Control: public, stale-while-invalidate`
2. `GET /`, hit
3. `PURGE /`
4. `GET /`, miss, refreshing
5. `GET /`, hit, serving stale until 4. resolves
6. request from 4. completes and stores new version in cache
7. `GET /`, hit

**Comment by IcedQuinn** [comment]
Came by to pass along an article I came across I thought might be useful here https://www.nginx.com/blog/benefits-of-microcaching-nginx/

Nginx has implemented an option to use a stale-while-invalidate which serves expired cache entries if multiple requests are hit while an attempt to update the cache is already in flight.

**Comment by darkweak** [comment]
Following the RFC, if the 5th request doesn't contains `stale-while-revalidate` or a `max-stale` directive we mustn't serve stale content. The only one thing we can do is to delay the fifth request while the fourth is revalidating and serve the fourth response content to the fifth one only if they have the same context (url, varied header values, ...).

**Comment by dkarlovi** [comment]
@darkweak the idea is you only allow this for responses which allow `stale-while-revalidate` indeed. Request 5 gets served the stale response created in request 1. The assumption is also the response will be refreshed long before `max-stale` is reached, but in that case indeed we must pass through as if the stale content doesn't even exist, it's basically evicted when `expires + max-stale` expire.

**Comment by dkarlovi** [comment]
The main problem here IMO is keeping track of the request 4. which is "the refresh request" here, its job is to refresh the response.

**Comment by darkweak** [comment]
@dkarlovi the next release will support that 👍
